File: ocaml/bin/day01.ml
let () = print_endline "Advent of Code 2024 day 1"

let part1 (list_left, list_right) =
  let sorted_left = List.sort compare list_left in
  let sorted_right = List.sort compare list_right in
  List.fold_left
    (fun sum x -> sum + x)
    0
    (List.map
       (fun (x, y) -> abs (x - y))
       (List.combine sorted_left sorted_right) )

let part2 (list_left, list_right) =
  let occur_left = Aoc2024.Utils.count_occurences list_left in
  List.fold_left
    (fun sum (value, count) ->
      sum
      + (count * value * Aoc2024.Utils.count_value_occurences value list_right)
      )
    0 occur_left

let solve input_filename =
  let lists = List.split (Aoc2024.Utils.read_list_of_pairs input_filename) in
  Stdio.printf "Part 1: Total distance = %d\n" @@ part1 lists ;
  Stdio.printf "Part 2: Similarity score = %d\n" @@ part2 lists

let () = solve "../inputs/day-01.txt"

File: haskell/lib/Day01.hs
module Day01 where

import Data.Bifunctor (Bifunctor (bimap))
import Data.List (sort)
import MyLib (countOccurences, getCounts, read2Cols)
import Text.Printf (printf)

distance :: (Num a, Ord a) => [a] -> [a] -> a
distance x y =
    sum $ abs <$> zipWith subtract (sort x) (sort y)

distance' :: (Num a, Ord a) => [a] -> [a] -> a
distance' x y = sum $ map abs $ uncurry (zipWith subtract) $ curry (bimap sort sort) x y

similarityScore :: (Num a, Eq a, Ord a) => [a] -> [a] -> a
similarityScore x y =
    sum $ (\(value, count) -> fromIntegral (count * countOccurences value y) * value) <$> getCounts x

part1 :: [Int] -> [Int] -> Int
part1 = distance

part2 :: [Int] -> [Int] -> Int
part2 = similarityScore

inputPath :: FilePath
inputPath = "../inputs/day-01-test.txt"

solution :: IO ()
solution =
    do
        putStrLn "Day 1"
        contents <- readFile inputPath
        let (col_left, col_right) = read2Cols contents :: ([Int], [Int])
        printf "Part 1: Distance = %d\n" $ part1 col_left col_right
        printf "Part 2: Similarity = %d\n" $ part2 col_left col_right
        print $ distance' col_left col_right
